### PR TITLE
feat(rules-package): enforce logical-key uniqueness + rename rp_manifests (CRD Issue 5a follow-up, #60)

### LINKED ISSUE #60: CRD Issue 5a follow-up — enforce logical-key uniqueness for RulesPackage and RuleSource
## Goal

Amend the Issue 5a Rules Package schema so that the logical identities and lifecycle constraints already assumed by the 5a/5b service layer are enforced cleanly at the DB/schema level.

## Why

Current service code treats these tuples as unique:
- `RulesPackage`: `(name, version, system)`
- `RuleSource`: `(rules_package_id, name)`

But the current ORM/schema does not enforce those logical keys. Under concurrent or repeated ingest paths, duplicate rows can be created, which can later cause `scalar_one_or_none()` lookups to fail or allow duplicate package/source identities to diverge.

This was surfaced during CRD Issue 5b review and is being handled as a schema-boundary follow-up rather than a silent 5b amendment.

Additional post-5b follow-ups surfaced during review and are intentionally folded into this issue so they have a canonical implementation home:

- **Package accessibility in `publish()` Gate 1:** `publish()` currently documents Gate 1 as requiring that the package "exists and is accessible," but the implementation only checks existence (`pkg_row is None`) and does not check accessibility state such as `is_enabled`. A disabled package could therefore still be marked published, creating contradictory lifecycle state. This was classified as non-blocking for 5b because the 5b acceptance criteria require publication gating on SQL ingest, manifest persistence, and vector write completeness — not package accessibility. This follow-up issue is the chosen place to decide and implement the correct lifecycle/schema behavior.

- **Manifest table naming consistency:** CRD Issue 5b explicitly required a dedicated `rules_package_manifests` table by name, so renaming it during 5b was correctly ruled out under issue supremacy. Now that 5b is complete, this follow-up issue is the approved place to perform that schema-family cleanup and align the manifest table with the `rp_*` naming convention if desired.

## In Scope

- Add DB uniqueness for `rp_packages` logical identity: `(name, version, system)`
- Add DB uniqueness for `rp_sources` logical identity: `(rules_package_id, name)`
- Remove/replace any redundant uniqueness constraint that does not protect the logical key
- Add/adjust Alembic migration(s)
- Add tests covering duplicate-insert rejection for both logical keys
- Review service upsert paths for conflict-safe behavior after the new constraints land
- Decide and implement the correct `publish()` Gate 1 behavior for package accessibility state (for example, whether disabled packages may be published)
- Update code/tests/docs so the Gate 1 contract and implementation agree
- Rename `rules_package_manifests` to align with the `rp_*` table family
- Add/update the migration(s), ORM mappings, service references, tests, and docs required by that manifest-table rename

## Out of Scope

- Any broader Rules Package schema redesign
- Changes to RuleChunk / MechanicalEntity provenance model
- Changes to Issue 5b ingestion semantics beyond what is needed to align with the new constraints and follow-up cleanup listed above

## Acceptance Criteria

- `rp_packages` enforces uniqueness on `(name, version, system)`
- `rp_sources` enforces uniqueness on `(rules_package_id, name)`
- Redundant uniqueness on `(rules_package_id, source_id)` is removed or justified
- Concurrent/duplicate logical inserts are rejected cleanly at the DB layer
- Tests cover both uniqueness constraints and service behavior against them
- `publish()` Gate 1 package-accessibility behavior is explicitly implemented and tested, or the contract is explicitly narrowed so it no longer over-claims accessibility checks
- The manifest table naming is standardized consistently across schema, migrations, ORM/service references, tests, and docs

### PR DESCRIPTION
**Pull request by ValravenAI**
## Summary

Closes #60 — CRD Issue 5a follow-up: enforce logical-key uniqueness for \`rp_packages\` and \`rp_sources\` at the DB layer, bring the manifest table into the \`rp_*\` naming family, and make both upsert helpers conflict-safe under the new constraints.

### Changes

- **\`rp_packages\`** — added \`UNIQUE(name, version, system)\` (\`uq_rp_packages_name_version_system\`): the logical identity key already enforced in read-before-insert by \`IngestionService._upsert_package()\`, now also enforced at the DB layer.
- **\`rp_sources\`** — added \`UNIQUE(rules_package_id, name)\` (\`uq_rp_sources_name\`): same pattern, mirrors \`IngestionService._upsert_source()\`. The existing \`uq_rp_sources_package_source\` constraint on \`(rules_package_id, source_id)\` is retained — it is required as the composite FK target for \`rp_chunks\` and \`rp_mechanical_entities\`.
- **\`rules_package_manifests\` → \`rp_manifests\`** — ORM tablename, constraint name, all runtime references, and the schema separation invariant test updated. Historical migration \`0004\` is untouched.
- **Migration \`0005\`** — uses \`op.batch_alter_table()\` for SQLite-compatible constraint additions; \`op.rename_table()\` for the manifest rename. Downgrade reverses all three operations.
- **\`publish()\` Gate 1** — added \`is_enabled\` check alongside the existence check. Publishing a disabled package would create contradictory lifecycle state (\`publication_status=published\` but invisible to play-time queries).
- **Conflict-safe upserts** — \`_upsert_package()\` and \`_upsert_source()\` now wrap the insert+flush in \`session.begin_nested()\`. If a concurrent writer inserts the same logical key between the read and the flush (raising \`IntegrityError\` on the new constraint), the savepoint is rolled back, the outer transaction is left intact, and the winner's row is re-read by logical key.
- **Tests** — new \`TestRpPackagesLogicalKeyUniqueness\` (4 tests), \`TestRpSourcesLogicalKeyUniqueness\` (3 tests), \`TestRpManifestTableName\` (2 tests), \`test_publication_fails_if_package_disabled\`, and \`TestUpsertConflictRecovery\` (2 tests covering the savepoint IntegrityError recovery path). Updated \`TestGetEntityByTypeAndName::test_scoped_to_package\` to use distinct logical-key values.

All 421 tests pass. Coverage: 96.44%.

## Architecture Notes

No drift from design principles.

- The new DB constraints mirror uniqueness guarantees already assumed by the service layer upsert paths — this tightens the schema to match existing behaviour, not changing semantics.
- \`uq_rp_sources_package_source\` is retained (not replaced) because it is a required composite FK target. Documented in both the ORM and the migration.
- The \`is_enabled\` check in \`publish()\` Gate 1 is within the issue boundary: the gate already checked package existence; checking enabled state is the natural complement and prevents contradictory lifecycle state.
- The \`rules_package_manifests\` → \`rp_manifests\` rename was explicitly listed in the issue scope as deferred from CRD Issue 5b.
- The savepoint-based concurrency fix is within scope: Issue #60 explicitly required reviewing service upsert paths for conflict-safe behaviour after the new constraints land.
- No Known Unknowns from \`/docs/architecture/known_unknowns.md\` are touched by this PR.

### CODE REVIEW

**Comment by ValravenAI** [comment]
## Afterworlds review classification summary

### Summary

PR #62 implements all six acceptance criteria from issue #60 cleanly: DB-layer uniqueness constraints for `rp_packages` and `rp_sources`, the `publish()` Gate 1 `is_enabled` guard, the `rules_package_manifests → rp_manifests` rename, migration `0005`, and a full test suite (96.42% coverage, 419 tests). Every change maps directly to an explicit AC in the approved issue text; no scope drift was found. Two items require resolution before merge: CI is still in progress at review time, and the PR description carries a merge-dependency warning whose accuracy needs owner confirmation before the queue advances.

---

### Merge-blocking defects

- **[DEFECT] CI gate has not yet completed.**
  - Why it blocks merge: Repository governance (`CLAUDE.md`, `construction_readiness.md` Item 12) prohibits merging any PR with failing or incomplete CI. The single check run ("Lint, Type-check, Test, Audit", run ID `71601153596`) was `in_progress` at review time (started 08:02:51Z, PR opened 08:02:44Z).
  - Evidence: `get_check_runs` response, `"status": "in_progress"`; `CLAUDE.md` rule "No PR merges with failing CI."

- **[DEFECT] Merge-dependency warning in PR description requires explicit owner confirmation before this PR advances.**
  - Why it blocks merge: The PR description states "This branch was cut from `feature/issue-5b-rules-package-ingestion-pipeline`, which is not yet merged to `main`." Migration `0005` has `down_revision = "0004"`, and `0004` is a CRD Issue 5b artifact (`0004_rules_package_manifests.py`, header: "Rules Package Manifests table — CRD Issue 5b"). If `0004` is absent from `main` when this PR merges, the Alembic chain is broken and all migration runs against `main` will fail at startup.
  - Mitigating evidence from the diff: `0004` does **not** appear in the PR diff, which normally means it is already present on the base branch (`main`, SHA `c7833e97`). This suggests the warning may be stale (5b may have already landed). However, the PR description itself says the opposite, and only the project owner can confirm the current state of `main`. The warning must be explicitly cleared or the base branch retargeted before this PR is eligible to merge.
  - Evidence: PR description "Merge dependency" block; `alembic/versions/0005_*.py` line `down_revision: Union[str, None] = "0004"`; `alembic/versions/0004_rules_package_manifests.py` header; PR diff (0004 absent → already on base, or stale warning — ambiguous without direct main inspection).

---

### Scope / boundary problems

None identified. All six change areas (rp_packages constraint, rp_sources constraint, manifest rename, migration 0005, Gate 1 `is_enabled` guard, tests) are explicitly listed in the issue #60 In Scope section and map to named acceptance criteria. The retention of `uq_rp_sources_package_source` satisfies the "removed **or justified**" AC — the justification (required composite FK target for `rp_chunks` and `rp_mechanical_entities`) is documented in both the ORM and the migration file.

---

### Known Unknowns

None touched. Confirmed against `/docs/architecture/known_unknowns.md` open items (React/Svelte, ChromaDB collection schema, FastAPI route shapes, rolling summary N value, session resumption UX, Writing mode persona details, prose parity). This PR is schema/ingestion-layer only and does not contact any of those items.

---

### Non-blocking improvements

- **[NON-BLOCKING] Constraint name `uq_rp_sources_name` understates the compound key it covers.**
  - The constraint enforces `(rules_package_id, name)`, not `name` alone. A name like `uq_rp_sources_package_name` would more precisely communicate the compound scope, consistent with how `uq_rp_packages_name_version_system` names its full key set. No runtime impact; the constraint functions correctly under any valid SQL identifier.
  - Why it should not block this PR: Naming is a style preference, not a correctness or schema-conformance defect. Renaming after the fact requires a new migration, which would be disproportionate churn.

- **[NON-BLOCKING] Inline imports inside test methods.**
  - `TestRpPackagesLogicalKeyUniqueness::test_upsert_package_finds_existing_row_after_constraint_added` and its `rp_sources` sibling import `IngestionService`, `IngestionConfig`, `datetime`, and `UTC` inside the test function body rather than at module level. The tests are correct and pass; the style is non-standard Python.
  - Why it should not block this PR: No correctness, coverage, or architectural impact. A follow-up cleanup in a later test-refactor pass is sufficient.

---

### Recommended next action

Wait for the CI run to complete. If it passes, confirm with the owner whether 5b has already landed on `main` (the diff evidence — `0004` absent from the diff — suggests it has, but the PR description says otherwise). If 5b is confirmed on `main`, update or remove the stale merge-dependency warning in the PR description and this PR is clear to merge. If 5b is **not** on `main`, retarget this PR to the 5b branch per the PR description's own instruction before merging.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/afterworlds/persistence/orm/rules_package.py [hunk 1]
```diff
@@ -29,6 +29,14 @@ class RulesPackageORM(Base):
     """Persisted RulesPackage row."""
 
     __tablename__ = "rp_packages"
+    __table_args__ = (
+        sa.UniqueConstraint(
+            "name",
+            "version",
+            "system",
+            name="uq_rp_packages_name_version_system",
```
[the remark this answers is not in the document]
Comment (ValravenAI) [reply src/afterworlds/persistence/orm/rules_package.py:37]: Addressed on the latest head commit.

`_upsert_package()` and `_upsert_source()` now wrap insert+flush in `session.begin_nested()` and recover from `IntegrityError` by re-reading the winner row by logical key, preserving the outer transaction. Added `TestUpsertConflictRecovery` coverage for both helpers.

The earlier PR merge-dependency note was stale and has been removed; `0004` is already on `main`.
